### PR TITLE
GCE PD plugin now prevents attaching a regional PD PV with pdName of …

### LINKED ISSUE #63317: PV for regional PD should not allow attaching of regular PDs
/kind bug

If a PV's zone annotation format indicates that it's a regional PD, but pdName corresponds to a regular PD, currently the GCE PD plugin allows attach to succeed. It should fail, because the mismatch is often caused by user error.

/sig gcp
/sig storage
/assign

### PR DESCRIPTION
**Pull request by verult**
…a regular PD

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #63317 

/sig storage
/sig gcp
/assign @msau42 @saad-ali

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/gce_disks.go [hunk 1]
```diff
@@ -535,14 +535,10 @@ func (gce *GCECloud) AttachDisk(diskName string, nodeName types.NodeName, readOn
 	if regional && utilfeature.DefaultFeatureGate.Enabled(features.GCERegionalPersistentDisk) {
 		disk, err = gce.getRegionalDiskByName(diskName)
 		if err != nil {
-			glog.V(5).Infof("Could not find regional PD named %q to Attach. Will look for a zonal PD", diskName)
-			err = nil
-		} else {
-			mc = newDiskMetricContextRegional("attach", gce.region)
+			return err
```
Comment (msau42) [inline pkg/cloudprovider/providers/gce/gce_disks.go]: can a unit test be added?

Comment (verult) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: Unfortunately I didn't have time to do the refactor for this release, and I'm planning to leave this unit test for either after code freeze (ideally) or the next release. Sorry this is non-ideal :(

File: pkg/cloudprovider/providers/gce/gce_disks.go [hunk 2]
```diff
@@ -535,14 +535,10 @@ func (gce *GCECloud) AttachDisk(diskName string, nodeName types.NodeName, readOn
...
 		}
-	}
-
-	if disk == nil {
```
Comment (saad-ali) [inline pkg/cloudprovider/providers/gce/gce_disks.go]: Why remove `if disk == nil {`?

Comment (verult) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: Before, if a regional PD is not found, we let the rest of the function continue to execute, so the `if disk == nil` is there to check whether a regional PD has been found.

Now, the function returns right away if an error is found, and `disk` is always nil if `gce.getDiskByName()` gets called

File: pkg/cloudprovider/providers/gce/gce_disks.go [hunk 3]
```diff
@@ -535,14 +535,10 @@ func (gce *GCECloud) AttachDisk(diskName string, nodeName types.NodeName, readOn
 	if regional && utilfeature.DefaultFeatureGate.Enabled(features.GCERegionalPersistentDisk) {
 		disk, err = gce.getRegionalDiskByName(diskName)
```
Comment (jingxu97) [inline pkg/cloudprovider/providers/gce/gce_disks.go]: what is regional is true, but feature is not enabled, or this case is not possible at all.

Comment (verult) [reply pkg/cloudprovider/providers/gce/gce_disks.go]: In that case we fall back to the behavior for a regular disk, as if the regional PD feature doesn't exist.

`regional` is set if the volume is a PV spec and its zone label has the regional PD format. If the feature gate is disabled, when a user creates a PV with regional PD zone label, the PV is rejected by an admission controller, so usually we don't hit this scenario.

**Comment by dims** [comment]
@verult needs a release note as well

**Pull request merged**